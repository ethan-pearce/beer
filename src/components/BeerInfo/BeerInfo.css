.beerInfo {
    margin: 32px 0;
    min-height: 600px;
}

.brewery {
    padding: 32px;
    border-bottom: 10px white;
    background-color: lightgrey;
}

.brewery:hover {
    opacity: 0.9;
}

.breweriesTitle {
    text-align: left;
    padding: 16px;
    border-top: 1px solid;
}

.location{
    margin: 16px;
    border: 1px solid;
    color: white;
}

.location:hover{
    margin: 16px;
    border: 1px solid black;
}

.locations{
    display: grid;
    grid-template-columns: repeat(4, 25%);
}

@media only screen and (max-width: 1023px) {
    .locations {
        display: grid;
        grid-template-columns: repeat(2, 50%);
    }
  }

.LocationTitle {
    text-align: left;
    padding:  0 16px;
}