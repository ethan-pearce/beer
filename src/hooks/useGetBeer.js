import { useState, useEffect } from 'react'
import { mapBeer } from '../Utils/mapBeer'

export const useGetBeer = (beerId) => {
    const [beer, setBeer] = useState([])
    const [loading, setLoading] = useState(true)
    const [error, setError] = useState(false)

    useEffect(() => {
        setLoading(true)

        fetch(`https://sandbox-api.brewerydb.com/v2/beer/${beerId}?withBreweries=Y&withSocialAccounts=Y&withIngredients=Y&key=816ecbe13a87ec829cd5aef047ca5c0a`)
            .then((res) => {
                if (!res.ok) {
                    setError(true)
                }
                return res.json()
            })
            .then((json) => {
                setBeer(mapBeerWithExtras(json.data))
                setLoading(false)
            })
            .catch(() => {
                setError(true)
                setLoading(false)
            })
    }, [beerId])

    return {
        beer,
        loading,
        error,
    }
}

const mapBeerWithExtras = (beer) => {
    return {
        ...mapBeer(beer),
        breweries: mapBreweries(beer.breweries)
    }
}

const mapBreweries = (breweries) => {
    const brews = breweries.map(brewery => {
        const { id, name, website, images, locations } = brewery;

        return {
            name,
            id,
            href: website,
            image: images.medium,
            locations: locations ? mapBreweryLocations(locations) : []
        }
    })

    return brews;
}

const mapBreweryLocations = (locations) => {
    const locat = locations.map(location => {
        const { id, name, locationTypeDisplay, region, postalCode } = location;

        return {
            name,
            id,
            type: locationTypeDisplay,
            region,
            postalCode
        }
    })

    return locat;
}