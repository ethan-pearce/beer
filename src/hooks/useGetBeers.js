import { useState, useEffect } from 'react'
import { mapBeer } from '../utils/mapBeer'

export const useGetBeers = () => {
    const [beers, setBeers] = useState([])
    const [loading, setLoading] = useState(true)
    const [error, setError] = useState(false)

    useEffect(() => {
        setLoading(true)

        fetch('https://sandbox-api.brewerydb.com/v2/beers?key=816ecbe13a87ec829cd5aef047ca5c0a')
            .then((res) => {
                if (!res.ok) {
                    setError(true)
                }
                return res.json()
            })
            .then((json) => {
                setBeers(mapBeers(json.data))
                setLoading(false)
            })
            .catch(() => {
                setError(true)
                setLoading(false)
            })
    }, [])

    return {
        beers,
        loading,
        error,
    }
}

const mapBeers = (beers) => {
    const mappedBeers = beers.map(beer => {
        return mapBeer(beer);
    })

    return mappedBeers;
}