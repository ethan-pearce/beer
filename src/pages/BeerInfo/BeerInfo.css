.beerInfo {
    margin: 0 0 32px 0;
    min-height: 600px;
}

.beerInfoTitle {
    margin: -32px 0 0 0;
    font-size: 116px;
}

@media only screen and (max-width: 1023px) {
    .beerInfoTitle {
        margin: -16px 0 0 0;
        font-size: 60px;
    }
}

.beerDescription{
    text-align: left;
    margin: 16px 16px 16px 8px;
}

@media only screen and (max-width: 1023px) {
    .beerDescription {
        max-height: 138px;
        overflow: scroll;
    }
}