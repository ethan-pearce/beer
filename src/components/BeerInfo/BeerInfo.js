import './BeerInfo.css';
import { useGetBeer } from '../../hooks/useGetBeer'
import { useParams } from "react-router-dom";
import BeerHeader from '../BeerHeader/BeerHeader'

const BeerInfo = () => {
  let { beerId } = useParams();
  const { beer, loading, error,} = useGetBeer(beerId)
  const { name, style, breweries, organic } = beer;

  return (
    <div className={'beerInfo'}>
        {loading && <div>Loading .....</div>}
        {error && <div>Sorry something went wrong</div>}
        {name &&
          <>
            <BeerHeader title={name} subTitle={style} organic={organic}/>
            {breweries.length > 0 &&
              <Breweries breweries={breweries}/>
            }
          </>
        }
    </div>
  );
}

export default BeerInfo;

const Breweries = ({breweries}) => {
  return (
    <>
      <h4 className={'breweriesTitle'}>Stocked at these breweries:</h4>
        <ul className={'breweries'}>
          {breweries.map(brewery => {
            const { href, name, image, locations, id } = brewery;
            return (
              <li key={id} className={'brewery'}>
                <a href={href}>
                  <h5>{name}</h5>
                  <p>{href}</p>
                  <img alt={name} src={image}/>
                </a>
                <Locations locations={locations}/>
              </li>
            )
          })}
      </ul>
    </>
  )
}

const Locations = ({locations}) => {
  if(locations.length <= 0){
    return null;
  }

  return (
    <>
      <p className={'LocationTitle'}>Locations:</p>
      <ul className={'locations'}>
        {locations.map(location => {
          const { id, name, type, region, postalCode } = location;
          return (
            <li key={id} className={'location'}>
              <p>{name}</p>
              <p>{`Type: ${type}`}</p>
              <p>{region}</p>
              <p>{postalCode}</p>
            </li>
          )
        })}
      </ul>
    </>
  )
}