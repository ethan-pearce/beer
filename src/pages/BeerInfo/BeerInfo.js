import React, { useContext } from 'react';
import './BeerInfo.css';
import { useGetBeer } from '../../hooks/useGetBeer'
import { useParams, Link } from "react-router-dom";
import BeerHeader from '../../components/BeerHeader/BeerHeader'
import BeerContext from '../../app/BeerContext'
import Loading from '../../components/Loading/Loading'

const BeerInfo = () => {
  let params = useParams();
  const { beer, loading, error } = useGetBeer(params?.beerId)
  const { breweries, description } = beer;
  const { beer: beerFromContext } = useContext(BeerContext);

  const name = beerFromContext.name || beer.name;
  const style = beerFromContext.style || beer.style;
  const organic = beerFromContext.organic || beer.organic;

  return (
    <div className={'beerInfo'}>
        {name ?
          <BeerHeader title={name} subTitle={style} organic={organic} titleClassName={'beerInfoTitle'}/>
        : null}
        {loading && <Loading />}
        {error && <div>Sorry something went wrong</div>}
        {description && <p className={'beerDescription'}>{description}</p>}
        {breweries &&
            breweries.length > 0 &&
              <Breweries breweries={beer.breweries}/>
        }
    </div>
  );
}

export default BeerInfo;

const Breweries = ({breweries}) => {
  return (
      <ul className={'breweries'}>
        {breweries.map(brewery => {
          const { href, name, image, locations, id } = brewery;
          return (
            <li key={id} className={'brewery'}>
              <Link to={`/brewery/${id}`} className={'breweryName'}>{name}</Link>
              <a className={'breweryLink'} href={href}>{href}</a>
              <img className={'breweryImage'} alt={name} src={image}/>
              <Locations locations={locations} breweryName={name}/>
            </li>
          )
        })}
    </ul>
  )
}

const Locations = ({locations, breweryName}) => {
  if(locations.length <= 0){
    return null;
  }

  return (
    <>
      <p className={'LocationTitle'}>Locations:</p>
      <ul className={'locations'}>
        {locations.map(location => {
          const { id, name, type, region, postalCode } = location;
          return (
            <li key={id} className={'location'}>
              <a href={`https://www.google.com/maps/search/?api=1&query=${breweryName}+${name}+${region}+${postalCode}`}>
                <p>{name}</p>
                <p>{`Type: ${type}`}</p>
                <p>{region}</p>
                <p>{postalCode}</p>
              </a>
            </li>
          )
        })}
      </ul>
    </>
  )
}