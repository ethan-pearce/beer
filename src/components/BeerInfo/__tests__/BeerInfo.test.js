import React from 'react'
import { render } from '@testing-library/react';
import BeerInfo from '../BeerInfo'
import { useParams } from "react-router-dom";
import { useGetBeer } from '../../../hooks/useGetBeer'

jest.mock('react-router-dom', () => ({
    useParams: jest.fn(),
}))

jest.mock('../../../hooks/useGetBeer', () => ({
    useGetBeer: jest.fn(),
}))

describe('BeerInfo Component', () => {
    it('should render when loading', () => {
        useParams.mockImplementation(() => { return { beerId: '43e33e2'}})
        useGetBeer.mockImplementation(() => { return { beer: {}, loading: true, error: false }})
        const { container } = render(<BeerInfo />)
        expect(container).toMatchSnapshot()
    })

    it('should render with error', () => {
        useParams.mockImplementation(() => { return { beerId: '43e33e2'}})
        useGetBeer.mockImplementation(() => { return { beer: {}, loading: false, error: true }})
        const { container } = render(<BeerInfo />)
        expect(container).toMatchSnapshot()
    })
})