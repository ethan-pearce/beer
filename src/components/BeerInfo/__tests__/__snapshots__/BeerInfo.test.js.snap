// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeerInfo Component should render when loading 1`] = `
<div>
  <div
    class="beerInfo"
  >
    <div>
      Loading .....
    </div>
  </div>
</div>
`;

exports[`BeerInfo Component should render with error 1`] = `
<div>
  <div
    class="beerInfo"
  >
    <div>
      Sorry something went wrong
    </div>
  </div>
</div>
`;
